File: processregistry.go
package gam

import "sync/atomic"

type ProcessRegistry struct {
	Node       string
	Host       string
	LocalPids  map[uint64]ActorRef
	SequenceID uint64
}

var GlobalProcessRegistry = ProcessRegistry{
	Node:      "nonnode",
	Host:      "nonhost",
	LocalPids: make(map[uint64]ActorRef),
}

func (pr ProcessRegistry) RegisterPID(actorRef ActorRef) *PID {
	id := atomic.AddUint64(&pr.SequenceID, 1)

	pid := PID{
		Node: pr.Node,
		Host: pr.Host,
		Id:   id,
	}

	pr.LocalPids[pid.Id] = actorRef
	return &pid
}

func (pr ProcessRegistry) FromPID(pid *PID) (ActorRef, bool) {
	if pid.Host != pr.Host || pid.Node != pr.Node {
		panic("Unknown host or node")
		return deadLetter, false
	}
	ref, ok := pr.LocalPids[pid.Id]
	if !ok {
		panic("Unknown PID")
		return deadLetter, false
	}
	return ref, true
}

File: languages/csharp/src/GAM/ProcessRegistry.cs
// -----------------------------------------------------------------------
//  <copyright file="ProcessRegistry.cs" company="Asynkron HB">
//      Copyright (C) 2015-2016 Asynkron HB All rights reserved
//  </copyright>
// -----------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Threading;

namespace GAM
{
    public class ProcessRegistry
    {
        private readonly IList<Func<PID, ActorRef>> _hostResolvers = new List<Func<PID, ActorRef>>();

        private readonly HashedConcurrentDictionary _localActorRefs =
            new HashedConcurrentDictionary();

        private int _sequenceId;
        public static ProcessRegistry Instance { get; } = new ProcessRegistry();

        public string Host { get; set; } = "nonhost";

        public void RegisterHostResolver(Func<PID, ActorRef> resolver)
        {
            _hostResolvers.Add(resolver);
        }

        public ActorRef Get(PID pid)
        {
            if (pid.Host != "nonhost" && pid.Host != Host)
            {
                foreach (var resolver in _hostResolvers)
                {
                    var reff = resolver(pid);
                    if (reff != null)
                    {
                        return reff;
                    }
                }
                throw new NotSupportedException("Unknown host");
            }

            ActorRef aref;
            if (_localActorRefs.TryGetValue(pid.Id, out aref))
            {
                return aref;
            }
            return DeadLetterActorRef.Instance;
        }

        public ValueTuple<PID, bool> TryAdd(string id, ActorRef aref)
        {
            var pid = new PID()
            {
                Id = id,
                Ref = aref, //cache aref lookup
                Host = "nonhost" // local
            };
            var ok = _localActorRefs.TryAdd(pid.Id, aref);
            return ValueTuple.Create(pid, ok);
        }

        public void Remove(PID pid)
        {
            _localActorRefs.Remove(pid.Id);
        }

        internal string GetAutoId()
        {
            var counter = Interlocked.Increment(ref _sequenceId);
            return "$" + counter;
        }
    }
}